File: typescript/src/web/AppFactory.ts
import SqlLitePostRepository from "../database/SqlLitePostRepository";
import SqlLiteUserRepository from "../database/SqlLiteUserRepository";
import PostRepository from "../domain/repositories/PostRepository";
import UserRepository from "../domain/repositories/UserRepository";

export default class AppFactory {

  private static userRepositoryInstance: UserRepository | null;
  private static postRepositoryInstance: PostRepository | null;

  static getUserRepository(): UserRepository {
    if (!this.userRepositoryInstance)
      this.userRepositoryInstance = new SqlLiteUserRepository('src/database/production.db')

    return this.userRepositoryInstance
  }

  static getPostRepository(): PostRepository {
    if (!this.postRepositoryInstance)
      this.postRepositoryInstance = new SqlLitePostRepository('src/database/production.db')

    return this.postRepositoryInstance
  }

  static resetRepositories(): void {
    this.userRepositoryInstance?.reset()
    this.postRepositoryInstance?.reset()
  }
}

File: java/src/main/java/tech/qmates/openchat/web/AppFactory.java
package tech.qmates.openchat.web;

import tech.qmates.openchat.database.SQLiteUserRepository;
import tech.qmates.openchat.domain.UTCClock;
import tech.qmates.openchat.domain.repository.UserRepository;

import java.net.URISyntaxException;
import java.net.URL;
import java.time.ZoneId;
import java.time.ZonedDateTime;

public class AppFactory {

    private static UserRepository userRepositoryInstance;

    public static UserRepository getUserRepository() {
        if (userRepositoryInstance == null)
            userRepositoryInstance = new SQLiteUserRepository(getSqliteFilePath());

        return userRepositoryInstance;
    }

    public static void resetRepositories() {
        if (userRepositoryInstance != null)
            userRepositoryInstance.reset();
    }

    public static UTCClock getRealClock() {
        return () -> ZonedDateTime.now(ZoneId.of("UTC"));
    }

    private static String getSqliteFilePath() {
        URL res = AppFactory.class.getClassLoader().getResource("production.db");
        if (res == null)
            throw new RuntimeException("Cannot find sqlite database file!");

        try {
            return res.toURI().getPath();
        } catch (URISyntaxException e) {
            throw new RuntimeException("Cannot load sqlite database file!", e);
        }
    }
}
